src/core/project/qgsprojectelevationproperties.cpp

```diff
@@ -108,7 +108,7 @@ QgsAbstractTerrainProvider *QgsProjectElevationProperties::terrainProvider()
 
 void QgsProjectElevationProperties::setTerrainProvider( QgsAbstractTerrainProvider *provider )
 {
-  if ( mTerrainProvider.get() == provider )
+  if ( mTerrainProvider.get()->equals( provider ) )
```

Comment:
This will introduce a leak now -- we'd need to check if it's `equals` but not `==` and if so, delete the passed object before returning.

Reply:
Indeed, good catch. Fixed.